src/Implementation/ColdStorage/ColdStorage/ColdStorageProcessor.cs

```diff
@@ -0,0 +1,327 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.ServiceModel.Channels;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Practices.IoTJourney.ColdStorage.Logging;
+using Microsoft.Practices.IoTJourney.ColdStorage.RollingBlobWriter;
+using Microsoft.Practices.IoTJourney.Devices.Events;
+using Microsoft.ServiceBus.Messaging;
+using Microsoft.WindowsAzure.Storage;
+using Newtonsoft.Json;
+
+namespace Microsoft.Practices.IoTJourney.ColdStorage
+{
+    public class ColdStorageProcessor : IEventProcessor
+    {
+        public const string EventDelimiter = "\r\n";
+        private const string ProcessorName = "writer";
+
+        private const int MaxBlocks = 5;
+        private const int MaxBlockSize = 4 * 1024 * 1024;
+
+        private readonly Func<string, IBlobWriter> _blobWriterFactory;
+        private readonly CancellationToken _token = CancellationToken.None;
+        private readonly int _maxBlockSize;
+
+        private readonly BufferManager _buffers;
+
+        private IBlobWriter _blobWriter;
+        private readonly List<BufferedFrameData> _eventHubBufferDataList = null;
+
+        private byte[] _currentFrame;
+        private int _currentFrameLength = 0;
+        private EventData _lastEventData;
+
+        // circuit breaker fields
+        private readonly int _warningLevel;
+        private readonly int _tripLevel;
+        private readonly TimeSpan _stallInterval;
+        private readonly TimeSpan _logCooldownInterval;
+        private readonly string _eventHubName;
+        private DateTime? _nextWarningLogTime;
+
+        public ColdStorageProcessor(
+            Func<string, IBlobWriter> blobWriterFactory,
+            CancellationToken token,
+            int warningLevel,
+            int tripLevel,
+            TimeSpan stallInterval,
+            TimeSpan logCooldownInterval,
+            string eventHubName,
+            int maxBlocks = MaxBlocks,
+            int maxBlockSize = MaxBlockSize)
+        {
+            Guard.ArgumentNotNull(blobWriterFactory, "blobWriterFactory");
```

Comment:
I'm also not sure about the `Guard` calls. 
I think these are useful for the surface area APIs, but we have them scattered liberally. 
How do you feel about them?